src/visitors.rs

```diff
@@ -679,21 +693,65 @@ impl<'a, 'b: 'a, 'tcx: 'b> MirVisitor<'a, 'b, 'tcx> {
                 .constant_value_cache
                 .check_if_std_intrinsics_unreachable_function(&fun)
             {
-                let span = self.current_span;
-                let mut err = self.session.struct_span_warn(
-                    span,
-                    "Control might reach a call to std::intrinsics::unreachable.",
-                );
-                (self.emit_diagnostic)(&mut err, &mut self.buffered_diagnostics);
+                //todo: use theorem prover to make sure that the entry condition is not known to be false.
+                if self
+                    .current_environment
+                    .entry_condition
+                    .as_bool_if_known()
+                    .unwrap_or(false)
```

Comment:
You have this .as_bool_if_known().unwrap_or(false) repeated so many times, it might be nice if there was a helper for it. Maybe you can have an OptionExt trait that adds:

  .as_bool_or_false()

It just feels a little nicer to me. I don't know, maybe it's not worth it. You can ponder it.

Reply:
It feels a little nicer to me too, but upon reflection the name is not satisfying. It also bears noting that the rather painful sequence it replaces has a desirable function in as much as it slows down the reader and forces them to remember that that logic is ternary, not binary.